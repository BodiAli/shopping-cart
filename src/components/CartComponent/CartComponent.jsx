import PropTypes from "prop-types";
import { useMemo, useState } from "react";
import Cart from "../../utils/cart";
import styles from "./CartComponent.module.css";

function CartComponent({
  cart,
  isActive,
  handleCartVisibility,
  handleInputChange,
  handleProductRemove,
  emptyCart,
}) {
  const [checkoutAnimation, setCheckoutAnimation] = useState(false);
  const [tickAnimation, setTickAnimation] = useState(false);

  const totalPrice = useMemo(() => {
    return cart.getAllPrices();
  }, [cart]);

  function handleClick() {
    setCheckoutAnimation(true);
  }

  let content;
  if (cart.count === 0) {
    content = <p className={styles.empty}>Cart is empty!</p>;
  } else {
    content = (
      <>
        {cart.products.map((product) => {
          return (
            <div key={product.title} className={styles.card}>
              <div className={styles.imageContainer}>
                <img
                  src={product.sample_cover ? product.sample_cover.thumbnail_image : ""}
                  alt={`${product.title} product poster`}
                />
              </div>
              <div className={styles.textContainer}>
                <h3>{product.title}</h3>
                <div className={styles.cartInfoContainer}>
                  <div className={styles.priceContainer}>
                    <p>{`Price: ${product.price}$`}</p>
                    <input
                      onChange={(e) => handleInputChange(e, product)}
                      value={product.quantity}
                      type="number"
                      name="quantity"
                      className={styles.quantityInput}
                    />
                  </div>
                  <button onClick={() => handleProductRemove(product)} className={styles.remove}>
                    Remove
                  </button>
                </div>
              </div>
            </div>
          );
        })}
        <button className={styles.checkout}>
          <p
            className={checkoutAnimation ? styles.checkoutAnimation : ""}
            onClick={handleClick}
            onAnimationEnd={() => {
              setTickAnimation(true);
            }}
          >
            Checkout
            <br />
            <span className={styles.total}>Total: {totalPrice}$</span>
          </p>

          <p
            className={`${styles.tick} ${tickAnimation ? styles.tickAnimation : ""}`}
            onAnimationEnd={() => {
              setTimeout(() => {
                setCheckoutAnimation(false);
                setTickAnimation(false);
                emptyCart();
              }, 2000);
            }}
          >
            Done!
          </p>
        </button>
      </>
    );
  }

  return (
    <section data-testid={"cart"} className={`${styles.cart} ${isActive ? styles.isActive : ""}`}>
      <button className={styles.closeButton} onClick={handleCartVisibility}>
        X
      </button>
      <div className={styles.container}>{content}</div>
    </section>
  );
}

CartComponent.propTypes = {
  cart: PropTypes.instanceOf(Cart).isRequired,
  isActive: PropTypes.bool,
  handleCartVisibility: PropTypes.func,
  handleInputChange: PropTypes.func.isRequired,
  handleProductRemove: PropTypes.func.isRequired,
  emptyCart: PropTypes.func,
};

export default CartComponent;
