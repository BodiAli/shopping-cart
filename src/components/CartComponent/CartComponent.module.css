.cart {
  position: absolute;
  background-color: #f0f0f0d8;
  top: 0;
  right: 0;
  left: 0;
  bottom: 0;
  z-index: 1;
  overflow-y: scroll;
  transform: translateX(100%);
  transition: transform 300ms ease;
  padding: 2rem;
}

.cart.isActive {
  transform: translateX(0);
}

.closeButton {
  position: absolute;
  top: 20px;
  right: 20px;
  background-color: var(--primary);
  border: none;
  border-radius: 5px;
  font-size: 1.8rem;
  font-weight: bold;
  padding: 0.5rem 1rem;
  cursor: pointer;
}

.closeButton:hover {
  filter: brightness(1.2);
}

.closeButton:active {
  filter: brightness(0.9);
}

.container {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  gap: 20px;
  margin: 0 250px;
  min-height: 100%;
}

.empty {
  font-family: "Merriweather", Times, serif;
  font-size: 3rem;
}

.card {
  border: 2px solid black;
  box-shadow: 2.8px 2.8px 10px rgba(0, 0, 0, 0.775);
  display: flex;
  margin: 0 200px;
  background-color: var(--gray);
  width: 100%;
}

.imageContainer {
  width: 30%;
}

.imageContainer > img {
  width: 100%;
  height: 200px;
  object-fit: contain;
}

.textContainer {
  padding: 1rem;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: start;
  width: 100%;
}

.textContainer > h3 {
  font-size: 2rem;
}

.priceContainer {
  display: flex;
  align-items: center;
  gap: 20px;
  box-shadow: 1px 1px 5px rgba(0, 0, 0, 0.775);
  padding: 0.5rem 1rem;
  border-radius: 5px;
}

.cartInfoContainer {
  display: flex;
  justify-content: space-between;
  align-items: center;
  width: 90%;
}

.remove {
  background-color: #c00000;
  border: none;
  border-radius: 5px;
  padding: 0.5rem 1rem;
  font-size: 1rem;
  cursor: pointer;
}

.remove:hover {
  filter: brightness(1.1);
}

.remove:active {
  filter: brightness(0.9);
}

.checkout {
  font-size: 1.7rem;
  padding: 0.5rem 1rem;
  border: none;
  border-radius: 5px;
  background: linear-gradient(270deg, #00c896, #7e7c7c);
  background-size: 400% 400%;
  font-weight: 600;
  margin-top: 25px;
  cursor: pointer;
  overflow: hidden;
  position: relative;
}

.checkout:hover {
  filter: brightness(1.1);
}

.checkout:active {
  filter: brightness(0.9);
}

.total {
  font-size: 1rem;
  color: var(--gray);
}

.checkoutAnimation {
  animation: scale-out 700ms ease 1 forwards;
}

@keyframes scale-out {
  from {
    transform: scale(1);
  }

  to {
    transform: scale(0);
  }
}

.tick {
  transform: translateX(-150%);
  position: absolute;
  top: 25%;
  left: 25%;
}

.tickAnimation {
  animation: move-in 500ms ease-out 1 forwards, shake 500ms ease 1;
}

@keyframes move-in {
  from {
    transform: translateX(-150%);
  }
  80% {
    transform: rotate(20deg);
  }
  to {
    transform: translateX(0);
  }
}
