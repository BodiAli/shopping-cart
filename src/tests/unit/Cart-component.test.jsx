import { render, screen } from "@testing-library/react";
import Cart from "../../utils/cart";
import CartComponent from "../../components/CartComponent/CartComponent";

const cart = new Cart([
  {
    title: "Fifa game",
    price: "50.50",
  },
  {
    title: "COD game",
    price: "50.50",
  },
  {
    title: "Minecraft game",
    price: "50.50",
  },
]);

describe("Cart Component", () => {
  test("should render correct elements", () => {
    const { container } = render(<CartComponent cart={cart} />);
    expect(container).toMatchSnapshot();
  });

  test("should render games in products array", () => {
    render(<CartComponent cart={cart} />);
    const games = screen.getAllByRole("heading", { name: /game$/i });
    games.forEach((game) => {
      expect(game).toBeInTheDocument();
    });
  });

  test("should render a remove button", () => {
    render(<CartComponent cart={cart} />);
    const removeButton = screen.getByRole("button", { name: /remove/i });
    expect(removeButton).toBeInTheDocument();
  });

  test("should have a checkout button", () => {
    render(<CartComponent cart={cart} />);
    const checkoutButton = screen.getByRole("button", { name: /checkout/i });
    expect(checkoutButton).toBeInTheDocument();
  });
});
