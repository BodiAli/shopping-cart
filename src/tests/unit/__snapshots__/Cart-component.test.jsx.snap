// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Cart Component > should render correct elements 1`] = `
<div>
  <section
    class="_cart_daa79f "
    data-testid="cart"
  >
    <button
      class="_closeButton_daa79f"
    >
      X
    </button>
    <div
      class="_container_daa79f"
    >
      <div
        class="_card_daa79f"
      >
        <div
          class="_imageContainer_daa79f"
        >
          <img
            alt="Fifa game product poster"
            src=""
          />
        </div>
        <div
          class="_textContainer_daa79f"
        >
          <h3>
            Fifa game
          </h3>
          <div
            class="_cartInfoContainer_daa79f"
          >
            <div
              class="_priceContainer_daa79f"
            >
              <p>
                Price: 50.50$
              </p>
              <input
                class="_quantityInput_daa79f"
                name="quantity"
                type="number"
                value="1"
              />
            </div>
            <button
              class="_remove_daa79f"
            >
              Remove
            </button>
          </div>
        </div>
      </div>
      <div
        class="_card_daa79f"
      >
        <div
          class="_imageContainer_daa79f"
        >
          <img
            alt="COD game product poster"
            src=""
          />
        </div>
        <div
          class="_textContainer_daa79f"
        >
          <h3>
            COD game
          </h3>
          <div
            class="_cartInfoContainer_daa79f"
          >
            <div
              class="_priceContainer_daa79f"
            >
              <p>
                Price: 50.50$
              </p>
              <input
                class="_quantityInput_daa79f"
                name="quantity"
                type="number"
                value="1"
              />
            </div>
            <button
              class="_remove_daa79f"
            >
              Remove
            </button>
          </div>
        </div>
      </div>
      <div
        class="_card_daa79f"
      >
        <div
          class="_imageContainer_daa79f"
        >
          <img
            alt="Minecraft game product poster"
            src=""
          />
        </div>
        <div
          class="_textContainer_daa79f"
        >
          <h3>
            Minecraft game
          </h3>
          <div
            class="_cartInfoContainer_daa79f"
          >
            <div
              class="_priceContainer_daa79f"
            >
              <p>
                Price: 50.50$
              </p>
              <input
                class="_quantityInput_daa79f"
                name="quantity"
                type="number"
                value="1"
              />
            </div>
            <button
              class="_remove_daa79f"
            >
              Remove
            </button>
          </div>
        </div>
      </div>
      <button
        class="_checkout_daa79f"
      >
        <p
          class=""
        >
          Checkout
          <br />
          <span
            class="_total_daa79f"
          >
            Total: 
            151.50
            $
          </span>
        </p>
        <p
          class="_tick_daa79f "
        >
          Done!
        </p>
      </button>
    </div>
  </section>
</div>
`;
